frame/glutton/src/lib.rs

```diff
@@ -0,0 +1,227 @@
+			let mut num_proof_size = 0;
+			while meter.can_accrue(
+				T::WeightInfo::waste_proof_size_some(1)
+					.max(T::WeightInfo::waste_proof_size_none(1)),
+			) {
+				let wasted = Self::waste_proof_size(num_proof_size);
+				num_proof_size.saturating_inc();
+				if wasted.is_zero() {
+					// Do not get stuck in an infinite loop if no PoV can be consumed.
+					break
+				}
+				if !meter.check_accrue(wasted) {
+					defensive!("Could not consume waste_proof_size");
+					break
+				}
+			}
+
+			// Now we waste ref time.
+			let mut num_ref_time = 0;
+			while meter.can_accrue(T::WeightInfo::waste_ref_time(1)) {
+				Self::waste_ref_time(num_ref_time);
+				num_ref_time.saturating_inc();
+				meter.defensive_saturating_accrue(T::WeightInfo::waste_ref_time(1));
+			}
+
+			meter.consumed
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Initializes the pallet by writing into `TrashData`.
+		///
+		/// Only callable by Root.
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::DbWeight::get().writes((*trash_count).into()))]
+		pub fn initialize_pallet(origin: OriginFor<T>, trash_count: u32) -> DispatchResult {
+			ensure_root(origin)?;
+
+			// Fill up the `TrashData` storage item.
+			(0..trash_count).for_each(|i| TrashData::<T>::insert(i, i));
+
+			Self::deposit_event(Event::PalletInitialized);
+			Ok(())
+		}
+
+		/// Set the `Compute` storage value that determines how much of the
+		/// block's weight `ref_time` to use during `on_idle`.
+		///
+		/// Only callable by Root.
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::DbWeight::get().writes(1))]
+		pub fn set_compute(origin: OriginFor<T>, compute: Perbill) -> DispatchResult {
+			ensure_root(origin)?;
+			Compute::<T>::set(compute);
+
+			Self::deposit_event(Event::ComputationLimitSet { compute });
+			Ok(())
+		}
+
+		/// Set the `Storage` storage value that determines the PoV size usage
+		/// for each block.
+		///
+		/// Only callable by Root.
+		#[pallet::call_index(2)]
+		#[pallet::weight(T::DbWeight::get().writes(1))]
+		pub fn set_storage(origin: OriginFor<T>, storage: Perbill) -> DispatchResult {
+			ensure_root(origin)?;
+			Storage::<T>::set(storage);
+
+			Self::deposit_event(Event::StorageLimitSet { storage });
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Wastes some `proof_size`. Receives a counter as an argument.
+		fn waste_proof_size(counter: u32) -> Weight {
+			if TrashData::<T>::get(counter).is_some() {
+				T::WeightInfo::waste_proof_size_some(1)
+			} else {
+				T::WeightInfo::waste_proof_size_none(1)
+			}
+		}
+
+		/// Wastes some `ref_time`. Receives a counter as an argument.
+		pub fn waste_ref_time(counter: u32) {
+			let mut hasher = Blake2b512::new();
+			// Blake2 has a very high speed of hashing so we make multiple
+			// hashes with it to waste more `ref_time` at once.
+			(0..50).for_each(|i: u32| {
+				hasher.update((counter & i).to_le_bytes());
+				let _ = hasher.clone().finalize();
+			});
+		}
```

Comment:
We need to use the result of `finalize`. Otherwise the compiler may optimizes this away. An idea could be that we or all values and then return it.:

```suggestion
		pub fn waste_ref_time(counter: u64) -> u64 {
			let mut hasher = Blake2b512::new();
			let mut result = 0;
			// Blake2 has a very high speed of hashing so we make multiple
			// hashes with it to waste more `ref_time` at once.
			(0..50).for_each(|i: u32| {
				hasher.update((counter & i).to_le_bytes());
				result |= hasher.clone().finalize();
			});
			
			result
		}
```

Then in the next round of calling `waste_ref_time` we take the previous output as input in `on_idle`. This should ensure that the compiler can not optimize it away.

Reply:
This function also shouldn't be public.

Reply:
Yeah, it makes sense to do something like you mentioned above, but that doesn't compile :P
I have changed the code so that the function doesn't receive a counter but instead it now receives a `Vec<u8>` which represents the previous result. This way the compiler shouldn't be able to do any optimization.